siatest/testgroup.go

```diff
@@ -0,0 +1,365 @@
+package siatest
+
+import (
+	"errors"
+	"math"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/node"
+	"github.com/NebulousLabs/Sia/node/api/client"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/fastrand"
+)
+
+type (
+	// GroupParams is a helper struct to make creating TestGroups easier.
+	GroupParams struct {
+		hosts   int // number of hosts to create
+		renters int // number of renters to create
+		miners  int // number of miners to create
+	}
+
+	// TestGroup is a group of of TestNodes that are funded, synced and ready
+	// for upload, download and mining depending on their configuration
+	TestGroup struct {
+		nodes   map[*TestNode]struct{}
+		hosts   map[*TestNode]struct{}
+		renters map[*TestNode]struct{}
+		miners  map[*TestNode]struct{}
+
+		dir string
+	}
+)
+
+var (
+	// defaultAllowance is the allowance used for the group's renters
+	defaultAllowance = modules.Allowance{
+		Funds:       types.SiacoinPrecision.Mul64(1e3),
+		Hosts:       5,
+		Period:      50,
+		RenewWindow: 10,
+	}
+)
+
+// addStorageFolderToHosts adds a single storage folder to each host.
+func addStorageFolderToHosts(hosts map[*TestNode]struct{}) error {
+	errors := make([]error, len(hosts))
+	wg := new(sync.WaitGroup)
+	i := 0
+	// The following api call is very slow. Using multiple threads speeds that
+	// process up a lot.
+	for host := range hosts {
+		wg.Add(1)
+		go func(i int, host *TestNode) {
+			errors[i] = host.HostStorageFoldersAddPost(host.Dir, 1048576)
+			wg.Done()
+		}(i, host)
+		i++
+	}
+	wg.Wait()
+	return build.ComposeErrors(errors...)
+}
+
+// announceHosts adds storage to each host and announces them to the group
+func announceHosts(hosts map[*TestNode]struct{}) error {
+	for host := range hosts {
+		if err := host.HostAcceptingContractsPost(true); err != nil {
+			return err
+		}
+		if err := host.HostAnnouncePost(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Close closes the group and all its nodes
+func (tg *TestGroup) Close() (err error) {
```

Comment:
All methods of a type should be grouped together.

Beyond that, we try to keep everything alphabetically sorted.